### ISSUE TITLE
Repository Permissions

### ISSUE THREAD

**Issue by willdean** [jakubgarfield/Bonobo-Git-Server#491]
At present, `IRepositoryRepository` has a method called `GetPermittedRepositories`, which is used by `RepositoryController` to get a list of repositories that the specified user is allowed to view - this is used to populate the index view (and @RedX2501's new pull-down list).

This method was semi-duplicated between AD and EF, but I combined them last week, so there's currently only one version of `GetPermittedRepositories`, used for both AD and EF.

Separately, there is a method on `IRepositoryPermissionService` called `HasPermission`, which takes a user Id and a Repo Id, and tells you if that user has permission on that specific repo.

Intuitively, you might think that `GetPermittedRepositories` would give the same answer as `GetAllRepositories().Where(...HasPermission(..))`.

However, this is not so - they differ in subtle ways:  

1. `GetPermittedRepositories` does not check to see if the specified user is the system admin - so this has to be done separately in `RepositoryController`.  However, it _will_ return all repos marked for anon access.

2. HasPermission **does** check to see if the specified user is system admin, but **doesn't** check if the repo is marked for anon access.  So when HasPermission is used, it tends to be coupled with a separate test for AllowsAnonymous

In addition to these rules being implemented multiple times in slightly different ways, the simple boolean concept of 'I have permission on this repo' is inadequate.   The rules for push and pull might be different for anonymous users, for example.  And permission to 'administer' repositories has to be checked separately.   This code ends up spread across multiple other places in the code, and the corner cases and interactions are very hard to check.   I came into the project because of bugs in the permissions logic, and there have been loads of issues around this sort of stuff, as well as feature requests which are hard to implement with confidence.  

I propose something like the following:

1. We define some levels of access request for a repository - I suggest 'Administer', 'Push' and 'Pull', but there might be something else we need.
2. We agree that Guid.Empty is the user ID of the anonymous user (I think it effectively is anyway)
3. We implement a single HasPermission method in the RepositoryService, which has a signature like this:
`bool HasPermission(Guid userId, Guid repositoryId, RepositoryAccessLevel level)`

Whenever you want to know if a user can access a repository, you call this routine, with the type of access you require.  (Mostly it would be 'Pull').   There's no function which returns an RepositoryAccessLevel in any way - it's just a level you ask for if you need.

Every kind of check for repo admin, system admin, team membership, anonymous rules, etc, _everything_ happens though this one call.   `GetPermittedRepositories` would then just be a helper which filtered the full list of repos through this call.

There are various requests like  #434 and #311 and even #77 which would be much easier if one could be more confident about this stuff, and issues like #390 would be much less likely to happen if the code wasn't so difficult work work on.

The nice think about pulling this code out of places like GitController, RepoController and the various auth attributes would be that it would be much easier to test for a consistent set of rules.

Thoughts?

tl;dr - I'd like to consolidate the currently scattered-about repo permission logic into a single point.

**Comment by willdean** [comment]
I think we need a 'Create' permission too.

**Comment by RedX2501** [comment]
I am absolutely in favour of having only one function that does permission checking.

I'm fine with the levels you propose.

Why do we need a `CreatePermission` function? Or is it supposed to be more like a `RepositoryAccessLevel GetPermissions(user_id, repository_id)`?

**Comment by willdean** [comment]
@RedX2501 The bit about 'Create' was me thinking that you need to be able to ask for permission to create a repository, so the levels would be 'Administer', 'Push', 'Pull' and 'Create'.

However, I've thought about this more since, and it doesn't make sense to have a method which takes a repositoryId guid which you ask for permission to create a new repository.

It would probably be a good idea to have a `bool CanCreateRepository(Guid userId)` method in the same class, but it wouldn't be the same method as I was proposing above.

I am 100% opposed to `RepositoryAccessLevel GetPermissions(user_id, repository_id)` because then people will start to combine RepositoryAccessLevel with other things, and put the RepositoryAccessLevel members in a particular order and compare them with `<` and  `>` and all sorts of other horrible malpractices.  My idea is just that you ask for what you need, and get told 'yes' or 'no'.

This is much the same way as some of the more modern authorisation tools work - see e.g. 'CheckAccess' here http://leastprivilege.com/2014/06/24/resourceaction-based-authorization-for-owin-and-mvc-and-web-api/

I'll stick up a PR and we can see what it looks like.

**Comment by RedX2501** [comment]
Ah ok.

Permissions are definitely not numbers and as such should not have mathematical operations applied to them. Only logical (maybe).

Looking forward for the PR.

**Comment by willdean** [comment]
> Permissions are definitely not numbers and as such should not have mathematical operations applied to them. Only logical (maybe).

I tend to try to imagine the worst possible perversion that fellow OSS authors could visit upon my carefully-crafted beautiful-to-behold code, then I'm not unduly disappointed when the reality turns out even worse :-)
